src/ImageSharp/Formats/Png/Filters/AverageFilter.cs

```diff
@@ -79,6 +84,89 @@ public static void Encode(Span<byte> scanline, Span<byte> previousScanline, Span
+                    hiRes32 = Avx2.UnpackHigh(hiRes16, Vector256<short>.Zero).AsInt32();
+                    sumAccumulator = Avx2.Add(sumAccumulator, loRes32);
+                    sumAccumulator = Avx2.Add(sumAccumulator, hiRes32);
+                }
+
+                for (int i = 0; i < Vector256<int>.Count; i++)
+                {
+                    sum += sumAccumulator.GetElement(i);
+                }
+            }
+            else if (Sse2.IsSupported)
+            {
+                var allBitsSet = Vector128.Create((sbyte)-1);
+                Vector128<int> sumAccumulator = Vector128<int>.Zero;
+
+                for (int xLeft = x - bytesPerPixel; x + Vector128<byte>.Count <= scanline.Length; xLeft += Vector128<byte>.Count)
+                {
+                    Vector128<byte> scan = Unsafe.As<byte, Vector128<byte>>(ref Unsafe.Add(ref scanBaseRef, x));
+                    Vector128<byte> left = Unsafe.As<byte, Vector128<byte>>(ref Unsafe.Add(ref scanBaseRef, xLeft));
+                    Vector128<byte> above = Unsafe.As<byte, Vector128<byte>>(ref Unsafe.Add(ref prevBaseRef, x));
+
+                    Vector128<byte> res = Sse2.Subtract(scan, Average(left, above));
+                    Unsafe.As<byte, Vector128<byte>>(ref Unsafe.Add(ref resultBaseRef, x + 1)) = res; // +1 to skip filter type
+                    x += Vector128<byte>.Count;
+
+                    Vector128<sbyte> absRes;
+                    if (Ssse3.IsSupported)
+                    {
+                        absRes = Ssse3.Abs(res.AsSByte()).AsSByte();
+                    }
+                    else
+                    {
+                        Vector128<sbyte> mask = Sse2.CompareGreaterThan(res.AsSByte(), Vector128<sbyte>.Zero);
+                        mask = Sse2.Xor(mask, allBitsSet);
+                        absRes = Sse2.Xor(Sse2.Add(res.AsSByte(), mask), mask);
+                    }
+
+                    Vector128<short> loRes16 = Sse2.UnpackLow(absRes, Vector128<sbyte>.Zero).AsInt16();
+                    Vector128<short> hiRes16 = Sse2.UnpackHigh(absRes, Vector128<sbyte>.Zero).AsInt16();
+
+                    Vector128<int> loRes32 = Sse2.UnpackLow(loRes16, Vector128<short>.Zero).AsInt32();
+                    Vector128<int> hiRes32 = Sse2.UnpackHigh(loRes16, Vector128<short>.Zero).AsInt32();
+                    sumAccumulator = Sse2.Add(sumAccumulator, loRes32);
+                    sumAccumulator = Sse2.Add(sumAccumulator, hiRes32);
+
+                    loRes32 = Sse2.UnpackLow(hiRes16, Vector128<short>.Zero).AsInt32();
+                    hiRes32 = Sse2.UnpackHigh(hiRes16, Vector128<short>.Zero).AsInt32();
+                    sumAccumulator = Sse2.Add(sumAccumulator, loRes32);
+                    sumAccumulator = Sse2.Add(sumAccumulator, hiRes32);
+                }
+
+                for (int i = 0; i < Vector128<int>.Count; i++)
+                {
+                    sum += sumAccumulator.GetElement(i);
+                }
```

Comment:
```suggestion
                if (Ssse3.IsSupported)
                {
                    Vector128<int> hadd = Ssse3.HorizontalAdd(sumAccumulator, sumAccumulator);
                    Vector128<int> swapped = Ssse3.Shuffle(hadd, 0x1);
                    Vector128<int> tmp = Sse2.Add(hadd, swapped);
                    sum += tmp.ToScalar();
                }
                else
                {
                    for (int i = 0; i < Vector128<int>.Count; i++)
                    {
                        sum += sumAccumulator.GetElement(i);
                    }
                }
```

Reply:
BTW: I'd write a helper `ReduceSum` (or similar) for this logic, so it can be re-used (and tested indepently). Same for AVX-version.

Reply:
Creating a helper would probably be the most sensible here, I agree.